### ISSUE TITLE
Enum codegen example

### ISSUE THREAD

**Issue by Yaneeve** [tminglei/slick-pg#171]
Thanks for your awesome work!

Could you please provide an example where I would generate code that would convert a postgres enum. I am new to scala and slick, and can't seem to get this right.

Thanks!

**Comment by tminglei** [comment]
Well, I'll update the code gen example project asap.

**Comment by Yaneeve** [comment]
Thanks! Appreciate it very much :)

**Comment by tminglei** [comment]
Hi @Yaneeve, I have submitted some updates in this [pull](https://github.com/tminglei/slick-pg/commit/e346bd72b65ea9999482328fdf64915ab49cd276), _but it's not ready yet._

**Comment by Yaneeve** [comment]
Yeah, I noticed... I certainly appreciate your responsiveness :) take as much time as you need

**Comment by tminglei** [comment]
Ok, I need more time. _;-)_

**Comment by bearrito** [comment]
I have working example if interested.

**Comment by tminglei** [comment]
@bearrito, yes, that's great. Can you send it to me?
_Thanks so much!_

**Comment by vrcod** [comment]
I can see this issue is still open? Is there a clear example somewhere of slick codegen working with Postgres enum types?

**Comment by tminglei** [comment]
Not yet.

But pg enum types can only be mapped to `String`, since it can't distinguish pg enum types from others and generate a scala type for them.

**Comment by vrcod** [comment]
Curiously, just to note something - JOOQ code generator works just fine with this situation - recognizes postgres enum types just fine and creates java enums accordingly, so it's not impossible, is it?

**Comment by tminglei** [comment]
@vrcod _thanks for the reminding! :-)_

I'll check JOOQ's codes to find how to distinguish and generate it.

**Comment by tlewowski** [comment]
just a follow up - I'd also love to see possibility of generating enumeration classes from database schema.

For my case (it's the simplest - just download all enum values together with labels and orders) sufficient SQL query was 

```sql
select t.typename, e.enumlabel, e.enumsortorder from pg_enum e 
  inner join pg_type t on t.oid=e.enumtypid
```
then some postprocessing is needed, but it generates just fine.

I should be able to post a full solution to GitHub later this week, let me know if you're interested

**Comment by tminglei** [comment]
@tlewowski yes, pls. 👍

**Comment by tlewowski** [comment]
I've posted a project here: https://github.com/tlewowski/pg-scala-enums

what it needs: 

 - adjustment of `dbUrl` in `build.sbt`
 - direct access to database for executing user (no user/password thing)

what it does:

 - fetches all enums via vanilla JDBC (no layers in between)
 - generates a file with all enumerations (as subclasses of `Enumeration`) - `gen-enums` task
 - generates a profile, based on `ExPostgresProfile` with many extensions (not configurable, changing extensions requires changes in code), with `implicit`s for generated enumerations included - `gen-pg-profile` task
 - generates a file with tables, where enumerations are included as normal types - `gen-tables`

stuff like package or file name is hardcoded in several places and naming convention for generated classes is pretty opinionated, but I don't think it'd be a big deal to change it if anybody needs it.

I've checked the mapping in my project (not included) and it works well with `PgEnumSupport`. 

BTW: thanks for this project, it really helps a lot!

**Comment by tminglei** [comment]
@tlewowski it looks good. :-)

**Comment by tlewowski** [comment]
well, the thing is that it doesn't fit very well into the current flow, so I'm not sure if it's mergeable anyhow, or should I maintain it as separate project (ideally an SBT plugin). any ideas?

**Comment by tminglei** [comment]
I think it should be kept as a separate project, since it's more like a solution instead of a patch.
